src/lib/Microsoft.Health.Common/Telemetry/MetricExtension.cs

```diff
@@ -38,7 +38,8 @@ public static Metric AddDimension(this Metric metric, string dimensionName, stri
                 return metric;
             }
 
-            metric.Dimensions.Add(dimensionName, dimensionValue);
+            metric.Dimensions[dimensionName] = dimensionValue;
```

Comment:
From the description of this bug, throwing the exception on duplicate dimensions seems like the correct thing to do. Otherwise the value would have been overridden which could have been hard to track down. 
If we really need the ability to override dimension names, I'd add in an optional parameter to this extension. Something like `overrideIfPresent`. And then do the override if the situation calls for it.

Reply:
If that is the direction might be good to switch to TryAdd and if the add fails that will be the trigger to throw.

Reply:
Throwing an exception in this case is also pretty difficult to track down since the error occurs at runtime instead of compile time, and in this case the error went through the FhirServiceExceptionProcessor which made the error pretty unclear. In the case of adding a duplicate metric key, an ArgumentException was thrown and then in [FhirServiceExceptionProcessor](https://github.com/microsoft/iomt-fhir/blob/c8f841610dab63c59519c14cdf7553b6085d6d48/src/lib/Microsoft.Health.Extensions.Fhir.R4/Telemetry/Exceptions/FhirServiceExceptionProcessor.cs#L62) there is a specific case for ArgumentException that expects a .ParamName which was null and so it threw a NullReferenceException and hid the original exception/ problematic code. 

Since the error also potentially affects message processing/retries it seemed that risk of overwriting the dimension value by default was safer than throwing an exception. For the PR I also just changed the dimension name from 'Name' to 'ResourceType' for the problematic metric so there wouldn't be a duplicate anymore, but wanted to put in some guardrails for the future because the error was really hard to track down today and there wasn't anything from stopping it from happening. We can remove it if we think it is a worse outcome to overwrite than to throw by default

Reply:
Maybe we can at least do an assert or some kind of trace so we have something we can track down if this ends up causing problems in the future and we need to track it down?

Reply:
Trace works for me.

Reply:
hmm I suppose a compromise would be to instead add a try/catch in .AddDimension with a Trace log whenever a dimension key that already exists is being added. However we don't have access to ITelemetryLogger in Microsoft.Health.Common where this MetricExtension lives. We could log to System.Console. Even though Console isn't ideal the error would at least be logged somewhere, the metric dimension would not be overwritten, and the message processing wouldn't be affected, which was one of my concerns. Any objections with that?

Reply:
I think if you use one of the methods in the System.Diagnostics it is better than a straight Console.WriteLine.  The System.Diagnostics can be easily configured/rerouted (even to one of our ILoggers)

Example https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.traceinformation?view=net-8.0

Reply:
thanks for the suggestion. updated .AddDimension to log an error trace when there is an error adding the dimension. the dimension is not added/overwritten